### ISSUE TITLE
Questions about sidecar

### ISSUE THREAD

**Issue by john-buildstar** [MotorsportReg/sidecar#8]
Greetings @ryanguill,

I've recently begun working with a large Coldfusion 11 project which I'm migrating from a physical server into Google's cloud (specifically, Google Container Engine, e.g., kubernetes). I'd like to scale the number of CF instances and therefore need a central location for session information.  Though sticky sessions work with Google's loadbalancer, code rollouts will interrupt them. Sidecar sounds ideal for my use case.  That said, I'm very new to Coldfusion so many things are presently unclear about what would be involved in using it.

Beyond this repo and your [introductory article on sidecar](http://ryanguill.com/cfml/2016/02/24/sidecar-external-session-management.html), I've not found evidence of anyone else using sidecar. You mention in your article that this code is used in production. 
* Is it still?
* Would you recommend its use, i.e., is it rock solid stable? Are there any significant gotchas?
* Do you know of anyone who has switched a very large application (which, presumably, had set sessions using, say, &lt;cfset session.key = val&gt;) to using it?
* Can you refer me to any additional information on installation and adapting an existing project to use it (especially given that my app uses components (.cfc) files very infrequently, if at all)?

In short, in your opinion, is the effort to use sidecar the best use of my time, or does it make more sense to upgrade to CF 2016 (which, I believe, also has an approach to shared, centralized state)?

Thanks much for any information you might provide.
John

**Comment by ghidinelli** [comment]
@atuttle just implemented in his app and could provide very recent experience if available.

**Comment by ryanguill** [comment]
@john-buildstar 

I think sidecar fits your use case perfectly (it is exactly the use case we designed it for), but it is not a drop in replacement for using the session scope, it will require you to update everywhere you use `session.` today.

We are not currently using it in production - shortly after we did we realized we had another component of the system that stores objects in the session that contain xml - and there is a bug in ACF that does not allow us to serialize objects containing xml.  So that is a limitation.

Also if you are using ORM there are potential limitations / problems.  I believe you can store the objects in sidecar and re-hydrate them fine, but they are not connected as they once were and I have not tested if its possible to get them reconnected without issues.

Ideally, in apps where you want this kind of functionality, you should store as little in your session as possible, and the things you store should be simple values - instead of storing objects store keys that you can go re-retrieve the object with.  That is a fine thing to do if you are designing an application with this in mind from the beginning, it can be harder when you are trying to retrofit an existing application.

Regarding stability - hopefully @atuttle can chime in and speak to his experience.  We do have a test suite that while it may not be 100% coverage, it is pretty good (and we would welcome any contributions on that front).  We know of no bugs or stability issues - but at the same time we make no guarantees.

I cannot speak to the features in ACF2016 - although I am skeptical it will be as robust of a solution as what we have - but again I have no basis for that except I know the challenges we came across when developing this.  

Re: my opinion on your scenario - I would evaluate how much effort it will be to update all places where you access and mutate the session scope today.  You will need to update any code that tries to set a value from `session.key = val` to `application.sess.set('key', val)` and anything that is reading from the session like `val = session.key` to `val = appliction.sess.get('key')`.  My understanding of the ACF functionality is that it should just be a configuration in the administrator and thats it, no code changes.

**Comment by john-buildstar** [comment]
Thanks, @ghidinelli, and thanks @ryanguill (for the detailed response),

Ryan, I don't follow what you said at the end of your post: "My understanding of the ACF functionality is that it should just be a configuration in the administrator and thats it, no code changes." Could you explain?  Thanks

**Comment by ryanguill** [comment]
I mean that I believe that the functionality in ACF2016 that you referred to would not require you to change how you access or mutate your session variables at all (no code changes), just would require you to set up the functionality in the administrator.  I expect it would involve configuring the redis instance you want to connect to - but thats all just conjecture, I haven't seen it at all.  You might find more information in the CFML slack group: cfml-slack.herokuapp.com

**Comment by john-buildstar** [comment]
I see; thanks.

**Comment by atuttle** [comment]
My understanding of the ACF2016 feature in question matches @ryanguill's explanation, but for what it's worth I haven't used it either. We're using Sidecar because we're migrating away from CFML and having session data portable between platforms was the first step in that direction. So it wouldn't make any sense for us to upgrade to ACF2016 to see it in action because odds of it being compatible with other platforms are (educated guess) nearly zero, and we don't want to pay to upgrade anyway.

Unfortunately my Sidecar implementation is not in production yet. We are currently planning to deploy it later this week (Thursday) and will be monitoring it closely when we do, but our expectation is that it will be a smooth transition. We do expect a _small_ increase in request time as a Redis read must be done at the start of each request, and a Redis write at the end, but so far tests are showing an addition of _much less than 100ms_, probably more like 20-30ms in most cases. Very acceptable considering this is a necessary step to migration without a full rewrite.

**Comment by john-buildstar** [comment]
@atuttle, how'd the deployment go?

**Comment by atuttle** [comment]
We only have our lowest-traffic production server on it for now, but so far everything seems to be functional and stable and performant. None of the users even noticed. We're planning to deploy to the rest of our production environments next week.

**Comment by john-buildstar** [comment]
Hello @ryanguill and @atuttle,

I've put sidecar in place in my development site. It generally works though I see a couple of different classes of problems that I'm starting to look at now. I'm wondering if you've encountered them before?

1) Exceeded 2 attempts to lock the resource redis-session-store_7AE6F509-A2BB-4259-168E04422DA8AD1C_lock

Exceeded 2 attempts to lock the resource redis-session-store_7AE6F509-A2BB-4259-168E04422DA8AD1C_lock
The error occurred in /site/prod_app/rev-db6f3ed3114a80fb60d3644d0b6f1bd04ec02070/ss/com/redis_session_store.cfc: line 46
Called from /site/prod_app/rev-db6f3ed3114a80fb60d3644d0b6f1bd04ec02070/ss/com/redlock.cfc: line 174
Called from /site/prod_app/rev-db6f3ed3114a80fb60d3644d0b6f1bd04ec02070/ss/com/redlock.cfc: line 87  

--- More lines cut ---

Is the expectation that anytime I set a value that I'll do a cflock?  I hope not because set I session vars in thousands of places :)

2) Performance is abysmal. I haven't yet begun to profile the code to see if there's some obvious wins. Given that redis should do tens of thousands of requests a second, I'd imagined that even if a given complex web page transaction reads and writes 500 times (which some do), that I'd be okay. 

3) I don't have a screen shot at the moment but I sometimes when doing a get, the md5sum of the session var is returned. I've added a couple of functions to work on nested structs (e.g., session.something.somekey = someval, etc) so it's possible that I've introduced this.

Anyway, any insight would be appreciated. Thanks, John

**Comment by atuttle** [comment]
We're now using Sidecar in production on our highest traffic sites and none of the end users have noticed -- performance has been great. Not sure if default sidecar works this way (I think it does) but my custom store only does 1 redis read at the start of the request and 1 write at the end of the request, so doing hundreds of writes in a request is no big deal. It still gets batched to 1 write at the end of the request.

I've not seen the errors you're posting. You might need to share some of your Application.cfc code.

**Comment by john-buildstar** [comment]
@atuttle, great to hear that you're using sidecar in production successfully. Well done! :)

It sounds like my app functions very differently from yours. For instance, many endpoints in my app result in 10-20 .cfm files being requested (it has a frameset structure, plus separate files for header elements, javascript libs (!!), etc.). Since all of them use session variables, I have an explosion of gets and sets. One test I did for the main screen showed about 5000 calls to sidecar; about 2/3 were cache hits, fortunately. If only I could share the connection that the first call (when sidecar.cfm is initialized) makes, that would probably help. (I see that sidecar uses a connection pool but I've not dug into it yet).

I'm starting on a test and benchmark suite today. Hoping it'll get me more clarity on where I'm falling down.

Does anyone have any recommendations on profiling Coldfusion?  Or, debugging for that matter. As a rookie Coldfusion guy, I'm relying on cfdump and sidecar's doLog().

thanks, John
